A27-xds-global-load-balancing.md

```diff
@@ -0,0 +1,440 @@
+policies, which is how the resolver and LB policies can share the same
+XdsClient object.
+
+#### CDS LB Policy
+
+The CDS LB policy will be the top-level LB policy.  Its configuration
+will be of the following form:
+
+```
+{
+  "cluster": "<cluster name>"
+}
+```
+
+The CDS policy will use the XdsClient object passed in from the `xds` resolver
+to query for the `Cluster` resource with the cluster name from the LB
+policy configuration.
+
+Initially, the CDS policy will always create an EDS policy as the child
+policy.  In the future, this may change; for example, we may add support
+for clusters that use DNS or for aggregate clusters, which would require
+different child policies.
+
+#### EDS LB Policy
+
+The EDS LB policy will be the child of the CDS LB policy.  Its
+configuration will be of the following form:
+
+```
+{
+  "edsServiceName": "<EDS service name>",
+  // Optional; if not specified, load reporting will be disabled.
+  // If value is the empty string, that means to use the xDS server as
+  // the LRS server.  (Currently, we do not support non-empty values;
+  // support for this will be added later.)
+  "lrsLoadReportingServerName": "<LRS server name>"
+}
+```
+
+The EDS policy will use the XdsClient object passed in from the xds
+resolver to query for the `ClusterLoadAssignment` resource with the
+EDS service name from the LB policy configuration.
+
+Note that unlike the old grpclb policy, which just received a flat list
+of backend servers from the balancer to round-robin across, the EDS policy 
+receives a hierarchical list, where the servers are grouped into
+localities, and the localities are grouped into priorities.  The EDS
+policy will be responsible for both selecting the highest priority set
+of localities that are reachable and distributing the traffic across the
+localities in the selected priority based on the locality weights.
+
+Note that the EDS policy will *not* support
+[overprovisioning](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/overprovisioning),
+which is different from Envoy.  Envoy takes the
+overprovisioning into account in both [locality-weighted load
+balancing](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/locality_weight)
+and [priority
+fail-over](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/priority),
+but gRPC assumes that the xDS server will update it to redirect traffic
+when this kind of graceful fail-over is needed.
+
+Unlike the grpclb policy, which relied on server-side load reporting to
+the balancer, the EDS policy will perform client-side load reporting
+using the [LRS
+protocol](https://github.com/envoyproxy/data-plane-api/blob/master/envoy/service/load_stats/v2/lrs.proto).
+Note that the EDS policy will not support per-endpoint stats; it will
+report only per-locality stats.  Currently, the EDS policy will report
```

Comment:
Do you mean that only the client-side stats are reported? These metrics do not seem to be useful for load balancing.

Reply:
I'm not sure which sentence you're asking about here.

The first sentence means that we will report only aggregate stats for the localities as a whole.  In particular, we will not support the [`report_endpoint_granularity` field](https://github.com/envoyproxy/envoy/blob/d25fb83ddb85d2de376df3fc91e15e6a3df02091/api/envoy/service/load_stats/v2/lrs.proto#L80) in the LRS response, nor will we populate the [`upstream_endpoint_stats` field](https://github.com/envoyproxy/envoy/blob/d25fb83ddb85d2de376df3fc91e15e6a3df02091/api/envoy/api/v2/endpoint/load_report.proto#L52) in the LRS load report.

The second sentence means is that we won't populate the [`load_metric_stats` field](https://github.com/envoyproxy/envoy/blob/d25fb83ddb85d2de376df3fc91e15e6a3df02091/api/envoy/api/v2/endpoint/load_report.proto#L47) in the LRS load reports, which is intended to contain data reported to the client by the backend.  The other fields, including number of requests sent/in flight/succeeded/failed, are all data that client already knows.

The data that we will report is sufficient for basic locality-level load balancing.  There are certainly more advanced scenarios that will need more functionality, but we'll add support for those later.

Reply:
What you mean is that in the second sentence, the client side will report these request information(e.g number of requests sent/in flight/succeeded/failed) to the manager server not the load of client itself?

Reply:
ok, I have got it. Thanks.